### PR TITLE
Basic structure

### PR DESCRIPTION
**Pull request by MarcCote**
This PR proposes a modular architecture for the new machine learning framework: `smartpy`. The primary blocs are: `Trainer`, `Optimizer`, `UpdateRule` and `Model`.

**Usage**
Training a model is simple as:
1. Select an `Optimizer` (e.g. `SGD`)
   - Set the loss function to optimize
   - Add one or several `UpdateRule`
2. Create a `Trainer`
   - Specify what model to train
   - Specify the dataset the model will be trained on
   - Specify the optimizer to use
3. Add tasks to the trainer 
   - e.g. saving the model each N epochs
   - e.g. early_stopping
   - e.g. printing the NLL after each update/epoch
   - e.g. printing the gradients
4. Add a viewer to the trainer
   - e.g. display filters
   - e.g. display reconstructed images
   - e.g. display gradient

**Discussion**
There are still some issues to be addressed/points to be discussed:
- [ ] What is the best approach to save/load models?
- [ ] What is the best approach to save/load training process?
- [ ] Is there a way to automatically generate argparse.Arguments for models
- [ ] What would be an ideal template for training scripts?

### CODE REVIEW

**Comment by MarcCote** [comment]
The architecture is not done yet but is now working enough to train a NADE model, see PR #2.

**Comment by MarcCote** [comment]
See simpler framework draft #3 .
